### ISSUE TITLE
Unable to open shell

### ISSUE THREAD

**Issue by roshanmohite** [F5Networks/f5-ansible#422]
##### ISSUE TYPE

NA

##### COMPONENT NAME

bigip_command

##### ANSIBLE VERSION

```
ansible 2.5.0
```

##### PYTHON VERSION

```
Python 2.7.6
```

##### BIGIP VERSION

```
Sys::Version
Main Package
  Product     BIG-IP
  Version     11.6.1
  Build       2.0.338
  Edition     Hotfix HF2
  Date        Fri Feb  3 14:44:37 PST 2017

```

##### CONFIGURATION

No changes made to ansible.cfg

##### SUMMARY

Trying to run commands using bigip_command module, however getting error "Unable to open shell".
I can login to LTM from Ansible host/controller using credentials specified in playbook. Gets me into      _admin@(nislab-f5v-rp01)(cfg-sync In Sync)(Active)(/Common)(tmos)#_ prompt.

##### STEPS TO REPRODUCE

```
---

- hosts: RC_F5
  gather_facts: true
  connection: local

  tasks:
   - name: run show version on remote devices
     bigip_command:
      transport: "cli"
      server_port: "22"
      validate_certs: "no"
      server: "{{ inventory_hostname}}"
#      server: "10.19.212.22"
      user: admin
      password: password
      commands: show sys clock
     delegate_to: localhost

```

##### EXPECTED RESULTS
Expecting to see Bigip system time.

##### ACTUAL RESULTS

```
nisadmin@NISLABL1:/etc/ansible/playbooks$ sudo ansible-playbook show_f5.yaml

PLAY [RC_F5] *****************************************************************************************************************************************************************

TASK [Gathering Facts] *******************************************************************************************************************************************************
ok: [UKCD_F5_1]
ok: [UKCD_F5_2]
ok: [UKCD_F5_4]
ok: [UKCD_F5_3]
ok: [UKLAB_F5]

TASK [run show version on remote devices] ************************************************************************************************************************************
fatal: [UKLAB_F5 -> localhost]: FAILED! => {"changed": false, "msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell"}
fatal: [UKCD_F5_2 -> localhost]: FAILED! => {"changed": false, "msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell"}
fatal: [UKCD_F5_1 -> localhost]: FAILED! => {"changed": false, "msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell"}
fatal: [UKCD_F5_4 -> localhost]: FAILED! => {"changed": false, "msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell"}
fatal: [UKCD_F5_3 -> localhost]: FAILED! => {"changed": false, "msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell"}
        to retry, use: --limit @/etc/ansible/playbooks/show_f5.retry

PLAY RECAP *******************************************************************************************************************************************************************
UKCD_F5_1                  : ok=1    changed=0    unreachable=0    failed=1
UKCD_F5_2                  : ok=1    changed=0    unreachable=0    failed=1
UKCD_F5_3                  : ok=1    changed=0    unreachable=0    failed=1
UKCD_F5_4                  : ok=1    changed=0    unreachable=0    failed=1
UKLAB_F5                   : ok=1    changed=0    unreachable=0    failed=1

nisadmin@NISLABL1:/etc/ansible/playbooks$
```

If I ssh to the LTM, it works fine, however i can't seem to run the script using the same login.

I am new to Ansible & Github, apologies if i've broken any hourse rules. 

Many thanks.

**Comment by roshanmohite** [comment]
Hello,
Further i made some changes to playbook as below:

![image](https://user-images.githubusercontent.com/28532002/32493510-1756b0a0-c3e4-11e7-8a81-bf7fb3a99254.png)

And it started to work. Below is my host file:
![image](https://user-images.githubusercontent.com/28532002/32493675-b30f4ffc-c3e4-11e7-82b7-f376cf870f71.png)

I cannot pin point what I am doing wrong with _server: "{{ inventory_hostname}}"_ attribute.

Any advice on this would be greatly appreciated. 
Thanks.

**Comment by caphrim007** [comment]
@roshanmohite there are several questions I need answered.

First, when you ssh to the box, are you specifying the user to SSH as? or are you just typing `ssh 1.1.1.1` and letting SSH decide. If the latter, then you appear to be going to the F5 device as the `nisadmin` user. Correct me if I'm wrong.

If you are, then you must __also__ specify the `user` value as `nisadmin` when you use the `transport: cli` setting. Here is how this works.

The `bigip_command` module has two modes. These modes are differentiated by the `transport` parameter. There is either `cli` or `rest`(the default).

Normally, when F5 Ansible modules run, they use REST to communicate with the BIG-IP. `bigip_command` is the **only** module, at this time, that allows you to change this behavior. You can change `bigip_command` to instead use SSH to communicate with the BIG-IP. Specifying `transport: cli` is how you tell the module to use SSH.

BIG-IPs have 2 administrative users; a CLI user named `root` and a web user named `admin` (this can differ in cloud environments but it is "generally" true).

When you specified `transport: cli` above, you also specified `user: admin` and this is __always__ incorrect unless you have specifically created an `admin` user that has CLI access. The `admin` user by default does not have CLI access (again...this differs in cloud environments).

When you then removed the `transport: cli` parameter, you caused the module to fallback to the default, REST, transport. Because there **is** a user named `admin` that exists in the REST API (as mentioned above `admin` is a web user and REST is a web transport) you saw success.

The general recommendation from me is to **NOT** use `transport: cli`. There are several reasons

- it is not supported by all modules
- it is "heavier" than REST; this can result is randomly dropped connections when using it; requiring extra `retries` be added to your playbooks
- finally, there is no such feature for other F5 products (BIG-IQ/Silverline, etc). So get used to using the REST API.

**Comment by roshanmohite** [comment]
Hello Tim,
With respect your 1st question about how am i logging into F5 LTM?
i do `> ssh admin@172.16.129.231 ` and then type in the password. I get log into 
`[root@nislab-f5v-rp01:Active:In Sync] ~ #` or `admin@(nislab-f5v-rp01)(cfg-sync In Sync)(Active)(/Common)(tmos)#` if i edit user privileges to use tmsh only. 
This playbook is run with 'admin' being able to access tmsh.

I will make sure not to change `transport` default value from REST. 

In the same playbook, I getting error if i user variable `"{{ inventory_hostname }}"` . 

```
---

- hosts: RC_F5
  gather_facts: true
  connection: local

  tasks:
   - name: run show version on remote devices
     bigip_command:
#      transport: "cli"
#      server_port: "22"
      validate_certs: "no"
      server: "{{ inventory_hostname}}"
#      server: "172.16.129.231"
      user: admin
      password: LTM$Adm!n
      commands: show sys clock
     delegate_to: localhost
     register: config

   - name: save output to /etc/ansible/backups
     copy:
       content: "{{ config.stdout[0] }}"
       dest: "/etc/ansible/backups/F5_{{ inventory_hostname }}_{{ansible_date_time.date}}.txt"

```

I get following error:

```
PLAY [RC_F5] *****************************************************************************************************************************************************************

TASK [Gathering Facts] *******************************************************************************************************************************************************
ok: [UKCD_F5_3]
ok: [UKCD_F5_4]

TASK [run show version on remote devices] ************************************************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: requests.exceptions.ConnectionError: HTTPSConnectionPool(host='UKCD_F5_4', port=443): Max retries exceeded with url: /mgmt/shared/authn/login (Caused by NewConnectionError('<requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x7f7316b24fd0>: Failed to establish a new connection: [Errno -2] Name or service not known',))
fatal: [UKCD_F5_4 -> localhost]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_UglYtY/ansible_module_bigip_command.py\", line 405, in <module>\n    main()\n  File \"/tmp/ansible_UglYtY/ansible_module_bigip_command.py\", line 390, in main\n    f5_product_name=spec.f5_product_name\n  File \"/tmp/ansible_UglYtY/ansible_modlib.zip/ansible/module_utils/f5_utils.py\", line 263, in __init__\n  File \"/tmp/ansible_UglYtY/ansible_modlib.zip/ansible/module_utils/f5_utils.py\", line 288, in _get_mgmt_root\n  File \"/usr/local/lib/python2.7/dist-packages/f5/bigip/__init__.py\", line 124, in __init__\n    self.post_configuration_setup()\n  File \"/usr/local/lib/python2.7/dist-packages/f5/bigip/__init__.py\", line 91, in post_configuration_setup\n    self._get_tmos_version()\n  File \"/usr/local/lib/python2.7/dist-packages/f5/bigip/__init__.py\", line 96, in _get_tmos_version\n    response = connect.get(base_uri)\n  File \"/usr/local/lib/python2.7/dist-packages/icontrol/session.py\", line 258, in wrapper\n    response = method(self, REST_uri, **kwargs)\n  File \"/usr/local/lib/python2.7/dist-packages/icontrol/session.py\", line 449, in get\n    return self.session.get(uri, **kwargs)\n  File \"/usr/local/lib/python2.7/dist-packages/requests/sessions.py\", line 487, in get\n    return self.request('GET', url, **kwargs)\n  File \"/usr/local/lib/python2.7/dist-packages/requests/sessions.py\", line 461, in request\n    prep = self.prepare_request(req)\n  File \"/usr/local/lib/python2.7/dist-packages/requests/sessions.py\", line 394, in prepare_request\n    hooks=merge_hooks(request.hooks, self.hooks),\n  File \"/usr/local/lib/python2.7/dist-packages/requests/models.py\", line 299, in prepare\n    self.prepare_auth(auth, url)\n  File \"/usr/local/lib/python2.7/dist-packages/requests/models.py\", line 494, in prepare_auth\n    r = auth(self)\n  File \"/usr/local/lib/python2.7/dist-packages/icontrol/authtoken.py\", line 207, in __call__\n    self.get_new_token(netloc)\n  File \"/usr/local/lib/python2.7/dist-packages/icontrol/authtoken.py\", line 104, in get_new_token\n    self.password))\n  File \"/usr/local/lib/python2.7/dist-packages/requests/api.py\", line 111, in post\n    return request('post', url, data=data, json=json, **kwargs)\n  File \"/usr/local/lib/python2.7/dist-packages/requests/api.py\", line 57, in request\n    return session.request(method=method, url=url, **kwargs)\n  File \"/usr/local/lib/python2.7/dist-packages/requests/sessions.py\", line 475, in request\n    resp = self.send(prep, **send_kwargs)\n  File \"/usr/local/lib/python2.7/dist-packages/requests/sessions.py\", line 585, in send\n    r = adapter.send(request, **kwargs)\n  File \"/usr/local/lib/python2.7/dist-packages/requests/adapters.py\", line 442, in send\n    raise ConnectionError(e, request=request)\nrequests.exceptions.ConnectionError: HTTPSConnectionPool(host='UKCD_F5_4', port=443): Max retries exceeded with url: /mgmt/shared/authn/login (Caused by NewConnectionError('<requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x7f7316b24fd0>: Failed to establish a new connection: [Errno -2] Name or service not known',))\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 0}
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: requests.exceptions.ConnectionError: HTTPSConnectionPool(host='UKCD_F5_3', port=443): Max retries exceeded with url: /mgmt/shared/authn/login (Caused by NewConnectionError('<requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x7f054351bfd0>: Failed to establish a new connection: [Errno -2] Name or service not known',))
fatal: [UKCD_F5_3 -> localhost]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_6OkI6e/ansible_module_bigip_command.py\", line 405, in <module>\n    main()\n  File \"/tmp/ansible_6OkI6e/ansible_module_bigip_command.py\", line 390, in main\n    f5_product_name=spec.f5_product_name\n  File \"/tmp/ansible_6OkI6e/ansible_modlib.zip/ansible/module_utils/f5_utils.py\", line 263, in __init__\n  File \"/tmp/ansible_6OkI6e/ansible_modlib.zip/ansible/module_utils/f5_utils.py\", line 288, in _get_mgmt_root\n  File \"/usr/local/lib/python2.7/dist-packages/f5/bigip/__init__.py\", line 124, in __init__\n    self.post_configuration_setup()\n  File \"/usr/local/lib/python2.7/dist-packages/f5/bigip/__init__.py\", line 91, in post_configuration_setup\n    self._get_tmos_version()\n  File \"/usr/local/lib/python2.7/dist-packages/f5/bigip/__init__.py\", line 96, in _get_tmos_version\n    response = connect.get(base_uri)\n  File \"/usr/local/lib/python2.7/dist-packages/icontrol/session.py\", line 258, in wrapper\n    response = method(self, REST_uri, **kwargs)\n  File \"/usr/local/lib/python2.7/dist-packages/icontrol/session.py\", line 449, in get\n    return self.session.get(uri, **kwargs)\n  File \"/usr/local/lib/python2.7/dist-packages/requests/sessions.py\", line 487, in get\n    return self.request('GET', url, **kwargs)\n  File \"/usr/local/lib/python2.7/dist-packages/requests/sessions.py\", line 461, in request\n    prep = self.prepare_request(req)\n  File \"/usr/local/lib/python2.7/dist-packages/requests/sessions.py\", line 394, in prepare_request\n    hooks=merge_hooks(request.hooks, self.hooks),\n  File \"/usr/local/lib/python2.7/dist-packages/requests/models.py\", line 299, in prepare\n    self.prepare_auth(auth, url)\n  File \"/usr/local/lib/python2.7/dist-packages/requests/models.py\", line 494, in prepare_auth\n    r = auth(self)\n  File \"/usr/local/lib/python2.7/dist-packages/icontrol/authtoken.py\", line 207, in __call__\n    self.get_new_token(netloc)\n  File \"/usr/local/lib/python2.7/dist-packages/icontrol/authtoken.py\", line 104, in get_new_token\n    self.password))\n  File \"/usr/local/lib/python2.7/dist-packages/requests/api.py\", line 111, in post\n    return request('post', url, data=data, json=json, **kwargs)\n  File \"/usr/local/lib/python2.7/dist-packages/requests/api.py\", line 57, in request\n    return session.request(method=method, url=url, **kwargs)\n  File \"/usr/local/lib/python2.7/dist-packages/requests/sessions.py\", line 475, in request\n    resp = self.send(prep, **send_kwargs)\n  File \"/usr/local/lib/python2.7/dist-packages/requests/sessions.py\", line 585, in send\n    r = adapter.send(request, **kwargs)\n  File \"/usr/local/lib/python2.7/dist-packages/requests/adapters.py\", line 442, in send\n    raise ConnectionError(e, request=request)\nrequests.exceptions.ConnectionError: HTTPSConnectionPool(host='UKCD_F5_3', port=443): Max retries exceeded with url: /mgmt/shared/authn/login (Caused by NewConnectionError('<requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x7f054351bfd0>: Failed to establish a new connection: [Errno -2] Name or service not known',))\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 0}
        to retry, use: --limit @/etc/ansible/playbooks/show_f5.retry

PLAY RECAP *******************************************************************************************************************************************************************
UKCD_F5_3                  : ok=1    changed=0    unreachable=0    failed=1
UKCD_F5_4                  : ok=1    changed=0    unreachable=0    failed=1
```
If i change user as `root` in play, it still throws above error.


Play runs fine when it is as below:
```
---

- hosts: RC_F5
  gather_facts: true
  connection: local

  tasks:
   - name: run show version on remote devices
     bigip_command:
#      transport: "cli"
#      server_port: "22"
      validate_certs: "no"
#      server: "{{ inventory_hostname}}"
      server: "172.16.129.231"
      user: admin
      password: LTM$Adm!n
      commands: show sys clock
     delegate_to: localhost
     register: config

   - name: save output to /etc/ansible/backups
     copy:
       content: "{{ config.stdout[0] }}"
       dest: "/etc/ansible/backups/F5_{{ inventory_hostname }}_{{ansible_date_time.date}}.txt"
 
```

**Comment by caphrim007** [comment]
@roshanmohite the error being reported gives you insight into what is wrong

`Failed to establish a new connection: [Errno -2] Name or service not known'`

This means that you are unable to resolve DNS. I bet your `{{ inventory_hostname }}` is a name, not an IP address, and that ansible is attempting to resolve this name in DNS.

Your play is telling me that your host's name is `UKCD_F5_4` and it's unlikely you have this name in eithre DNS or your /etc/hosts file.

**Comment by roshanmohite** [comment]
Hey Tim,
My hosts file looks like below:
```
[RC_F5]
#UKLAB_F5 ansible_host=10.19.212.22
#UKCD_F5_1 ansible_host=172.16.130.29
#UKCD_F5_2 ansible_host=172.16.130.30
UKCD_F5_3 ansible_host=172.16.129.231
UKCD_F5_4 ansible_host=172.16.129.232

```
I am using  `{{ inventory_hostname }}` hoping it pick IP from hosts file. 
While looking at the host file however, I thought to use `ansible_host` instead of `inventory_hostname`; this seems to have done the trick and my play is running as expected.
Now the playbook looks as below:
```
---

- hosts: RC_F5
  gather_facts: true
  connection: local

  tasks:
   - name: run show version on remote devices
     bigip_command:
#      transport: "cli"
#      server_port: "22"
      validate_certs: "no"
      server: "{{ ansible_host }}"
#      server: "172.16.129.231"
      user: admin
      password: LTM$Adm!n
      commands: show sys clock
     delegate_to: localhost
     register: config

   - name: save output to /etc/ansible/backups
     copy:
       content: "{{ config.stdout[0] }}"
       dest: "/etc/ansible/backups/F5_{{ inventory_hostname }}_{{ansible_date_time.date}}.txt"

```

Does it mean when `inventory_hostname` is passed as variable to `host:` it picks up hostname from hosts file & does a DNS look up on hostname defined in hosts file; on the other hand `ansible_host` picks up the IP defined in hosts file?

Many thanks

**Comment by caphrim007** [comment]
@roshanmohite you're experiencing the difference between two different variables with two different values and how they are influenced by your network.

Here is a line from your inventory

`UKCD_F5_4 ansible_host=172.16.129.232`

In this line, there are two values. One of them is implicitly defined; the other explicitly defined. They are

* `inventory_hostname`. Value, `UKCD_F5_4`. Implicitly defined
* `ansible_host`. Value, `172.16.129.232`. Explicitly defined

In your playbooks, you were referencing `inventory_hostname`. In many large production networks, this works just fine because the hosts that are in inventory usually are also dns names. For example

```
google.com
microsoft.com
mit.edu
```

Those could all be put in a hosts file, would be read in as `inventory_hostname` and would work just fine when it came time for Ansible to connect to them because they are resolvable hosts on a network.

But you've put a hostname that presumably has no entry in either DNS or your /etc/hosts file

`UKCD_F5_4`

Therefore, when you go to use it in your playbook, ansible tries to connect to it via SSH, name resolution on your PC attempts to happen, and your DNS server returns nothing because it knows of no host named `UKCD_F5_4`.

Then you changed your playbook to use `ansible_host`.

Again, ansible tries to connect to it via SSH, no name resolution needs to happen because it is an IP address, TCP connection is made to the remote device because it is online, and the task finished without issue.

The bottom line is that you were attempting to use ansible examples in an environment that does not have DNS configured for your hosts.

If you wanted to continue to use `inventory_hostname`, then the proper way to do that is to just specify IP addresses in your hosts file

```
192.168.1.2
192.168.1.3
192.168.1.4
```

And then Ansible will use IP addresses instead of names.

**Comment by roshanmohite** [comment]
Understood. Many thanks!
